core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -20,33 +20,143 @@
+   */
+  @VisibleForTesting
+  Optional<Duration> cachedEntryAge(TableIdentifier identifier) {
+    return tableCache.policy()
+        .expireAfterAccess()
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier));
+  }
+
+  // Returns the cached Table entry corresponding to the given identifier iff
+  // it's in the cache. Grabs the table in a way that doesn't count as an access
+  // and thus won't affect the cached entry's ttl (if enabled).
+  @VisibleForTesting
+  Optional<Table> tableFromCacheQuietly(TableIdentifier identifier) {
+    // Ensure async cleanup actions have happened.
+    tableCache.cleanUp();
+    return Optional.ofNullable(tableCache.policy().getIfPresentQuietly(identifier));
+  }
+
+  @VisibleForTesting
+  Cache<TableIdentifier, Table> cache() {
+    return tableCache;
+  }
+
+  @VisibleForTesting
+  Optional<Duration> getTimeToTTL(TableIdentifier identifier) {
+    return tableCache
+        .policy()
+        .expireAfterAccess()  // Assumes expireAfterAccess, which is what we set at cache level.
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier)) // Get the time the table has been cached.
+        .map(age -> Duration.ofMillis(expirationIntervalMillis).minus(age));
+  }
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<TableIdentifier, Table> cacheBuilder = Caffeine
+        .newBuilder()
+        .softValues()
+        .removalListener(keyLoggingRemovalListener)
+        .writer(new CacheWriter<TableIdentifier, Table>() {
+          @Override
+          // TODO - Consider expiring and syncing any metadata tables that have a different snapshotId
+          //        upon write.
+          public void write(TableIdentifier tableIdentifier, Table table) {
+            LOG.info("Table {} was written to the catalog with snapshotId {}", tableIdentifier,
+                table.currentSnapshot() == null ? null : table.currentSnapshot().snapshotId());
+          }
+
+          @Override
+          public void delete(TableIdentifier tableIdentifier, Table table, RemovalCause cause) {
+            // On expiration, remove any associated metadata tables so that subsequent catalog loads won't
+            // return stale metadata tables w.r.t. the underlying data tables they would return.
+            //
+            // TODO - Should we put metadata tables back into the catalog if their associated table is still
+            //        cached to keep tables and metadata tables on the same snapshot?
+            if (expirationEnabled && !MetadataTableUtils.hasMetadataTableName(tableIdentifier)) {
+              onTableExpiration(tableIdentifier);
+            }
+          }
+        });
+
+    if (expirationEnabled) {
+      // Expire after write isn't going to work as we need to invalidate after writes.
+      return cacheBuilder.expireAfterAccess(Duration.ofMillis(expirationIntervalMillis)).ticker(ticker).build();
+    }
+    return cacheBuilder.build();
```

Comment:
Nit: newline between control flow and return.

Reply:
Added.